File: 03-11-2024_Rotate_string.java
class Solution {
    public boolean rotateString(String s, String goal) {
        if (s.length() != goal.length()) {
            return false;
        }

        s = s + s;
        int i = 0, j = 0;
        int n = goal.length();

        while (i < 2 * n && j < n) {
            if (s.charAt(i) == goal.charAt(j)) {
                i++;
                j++;
            } else {
                i = i - j + 1;
                j = 0;
            }

            if (j == n) {
                return true;
            }
        }

        return false;
    }
}

File: 03-11-2024_Rotate_string.py
class Solution:
    def rotateString(self, s: str, goal: str) -> bool:
        if len(s) != len(goal):
            return False

        s = s + s
        i, j = 0, 0
        n = len(goal)

        while i < 2 * n and j < n:
            if s[i] == goal[j]:
                i += 1
                j += 1
            else:
                i = i - j + 1
                j = 0
            
            if j == n:
                return True

        return False
